File: apps/desktop/updater.js
// 앱 업데이터 — 서버 매니페스트(GET /api/desktop/latest) 기반 자체 교체.
//
// 미서명(ad-hoc) 배포라 Squirrel/electron-updater 를 쓸 수 없어(서명 필수),
// dmg 를 받아 sha256 검증 후 분리 프로세스가 /Applications 교체 + 재실행한다.
// 흐름: 버전 비교 → 사용자 확인 다이얼로그 → 다운로드(tmp) → sha256 검증 →
//       교체 스크립트 spawn(detached) → 앱 종료 → 스크립트가 교체·xattr·재실행.
const { app, dialog, shell } = require('electron');
const fs = require('fs');
const os = require('os');
const path = require('path');
const crypto = require('crypto');
const { spawn } = require('child_process');

const CHECK_DELAY_MS = 5000;
// 매니페스트 파일명 화이트리스트 — 교체 스크립트에 문자열로 보간되므로 서버(config/
// desktop-update.ts)와 같은 패턴을 앱에서도 검증한다(백엔드 전환 가능 → 서버 검증만 의존 불가).
const FILE_PATTERN = /^OpenMake-[A-Za-z0-9.-]+\.dmg$/;
const SHA256_PATTERN = /^[0-9a-f]{64}$/i;

// 업데이트는 미서명(ad-hoc) 교체라 무결성 신뢰가 전송 보안에 달려 있다. HTTPS 또는
// 루프백(로컬 백엔드)에서만 허용해 평문 HTTP MITM 으로 악성 dmg 가 주입되는 것을 막는다.
function isSecureUpdateOrigin(backendUrl) {
  try {
    const u = new URL(backendUrl);
    if (u.protocol === 'https:') return true;
    return ['localhost', '127.0.0.1', '::1', '[::1]'].includes(u.hostname);
  } catch { return false; }
}

/** "1.2.0" 형태 비교 — a > b 이면 양수. */
function cmpVersion(a, b) {
  const pa = String(a).split('.').map(Number);
  const pb = String(b).split('.').map(Number);
  for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
    const d = (pa[i] || 0) - (pb[i] || 0);
    if (d) return d;
  }
  return 0;
}

function fetchJson(url) {
  return fetch(url, { headers: { Accept: 'application/json' } }).then((r) => {
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    return r.json();
  });
}

async function downloadTo(url, dest) {
  const r = await fetch(url);
  if (!r.ok) throw new Error(`다운로드 실패 HTTP ${r.status}`);
  const buf = Buffer.from(await r.arrayBuffer());
  fs.writeFileSync(dest, buf);
  return crypto.createHash('sha256').update(buf).digest('hex');
}

/** 교체 스크립트 — 앱 종료 후 dmg 마운트 → /Applications 교체 → 격리 해제 → 재실행. */
function spawnReplaceScript(dmgPath) {
  const script = `#!/bin/bash
sleep 2
MNT=$(hdiutil attach -nobrowse "${dmgPath}" | tail -1 | awk '{print $NF}')
if [ -d "$MNT/OpenMake.app" ]; then
  rm -rf /Applications/OpenMake.app
  ditto "$MNT/OpenMake.app" /Applications/OpenMake.app
  xattr -dr com.apple.quarantine /Applications/OpenMake.app 2>/dev/null
fi
hdiutil detach "$MNT" -quiet
rm -f "${dmgPath}"
open -a /Applications/OpenMake.app
rm -f "$0"
`;
  const sh = path.join(os.tmpdir(), `openmake-update-${Date.now()}.sh`);
  fs.writeFileSync(sh, script, { mode: 0o755 });
  spawn('/bin/bash', [sh], { detached: true, stdio: 'ignore' }).unref();
}

/**
 * 업데이트 확인. interactive=true(메뉴)면 "최신 버전" 안내도 표시,
 * false(기동 시 자동)면 새 버전 있을 때만 다이얼로그.
 */
async function checkForUpdate(backendUrl, win, interactive) {
  try {
    if (!isSecureUpdateOrigin(backendUrl)) {
      throw new Error('업데이트는 HTTPS(또는 로컬호스트) 백엔드에서만 허용됩니다');
    }
    const m = (await fetchJson(`${backendUrl}/api/desktop/latest`)).data;
    // sha256 은 필수 — 없으면(또는 형식 불량) 무결성 검증이 불가하므로 거부한다.
    if (!m || !m.version || !FILE_PATTERN.test(String(m.file)) || !SHA256_PATTERN.test(String(m.sha256 || ''))) {
      throw new Error('업데이트 매니페스트 형식이 올바르지 않습니다(version·file·sha256 확인)');
    }
    const current = app.getVersion();
    if (cmpVersion(m.version, current) <= 0) {
      if (interactive) {
        dialog.showMessageBox(win, { type: 'info', message: `최신 버전입니다 (v${current})` });
      }
      return;
    }
    const autoTest = process.env.OMK_UPDATE_AUTO === '1'; // 테스트 훅(개발 전용): 다이얼로그 생략
    const choice = autoTest ? 0 : dialog.showMessageBoxSync(win, {
      type: 'info',
      message: `새 버전 v${m.version} 이 있습니다 (현재 v${current})`,
      detail: '지금 업데이트하면 앱이 잠시 종료된 뒤 새 버전으로 다시 열립니다.',
      buttons: ['지금 업데이트', '나중에'],
      defaultId: 0,
      cancelId: 1,
    });
    if (choice !== 0) return;

    const dmgPath = path.join(os.tmpdir(), m.file);
    const sha = await downloadTo(`${backendUrl}${m.url}`, dmgPath);
    if (sha.toLowerCase() !== String(m.sha256).toLowerCase()) {
      fs.rmSync(dmgPath, { force: true });
      throw new Error('다운로드 무결성 검증 실패(sha256 불일치)');
    }
    spawnReplaceScript(dmgPath);
    app.quit();
  } catch (e) {
    if (interactive) {
      dialog.showMessageBox(win, {
        type: 'warning',
        message: '업데이트 확인 실패',
        detail: String(e.message || e),
      });
    }
    // 자동 확인 실패는 조용히 무시 (오프라인 등)
  }
}

module.exports = {
  checkForUpdate,
  /** 기동 시 자동 확인(지연 실행) — 새 버전 있을 때만 표시. */
  scheduleStartupCheck: (backendUrlFn, winFn) => {
    setTimeout(() => checkForUpdate(backendUrlFn(), winFn(), false), CHECK_DELAY_MS);
  },
  openReleasePage: (backendUrl) => shell.openExternal(backendUrl),
};

File: apps/desktop-native/OpenMakeCompanion/Sources/OpenMakeCompanion/Updater.swift
// 자체 업데이터 — 서버 매니페스트(native 채널) 확인 → dmg 다운로드 → sha256 검증 →
// 분리 프로세스가 앱 교체 + 재실행 (Electron updater.js 와 동일 흐름·동일 안전장치).
//
// 보안:
//  - 업데이트 확인은 HTTPS(또는 로컬호스트) 백엔드에서만 허용 (MITM 차단)
//  - sha256 필수 — 매니페스트에 없으면 거부 (ad-hoc 서명 배포의 무결성 검증 수단)
//  - 교체는 스테이징(.new) → 스왑 → 실패 시 롤백(.old) — 도중 실패에도 기존 앱 보존
import AppKit
import CryptoKit
import Foundation

@MainActor
final class Updater: ObservableObject {
    static let shared = Updater()

    @Published var checking = false

    private var currentVersion: String {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0"
    }

    /** 기동 5초 후 자동 확인 (Electron 관행 준수). 새 버전 있을 때만 다이얼로그. */
    func scheduleStartupCheck(backendUrl: String) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
            Task { await self.check(backendUrl: backendUrl, interactive: false) }
        }
    }

    func check(backendUrl: String, interactive: Bool) async {
        guard !checking else { return }
        checking = true
        defer { checking = false }
        do {
            guard isSecureOrigin(backendUrl) else {
                throw UpdateError.message(L("update.insecure"))
            }
            guard let url = URL(string: "\(backendUrl)/api/desktop/latest") else {
                throw UpdateError.message(L("update.badBackend"))
            }
            let (data, _) = try await URLSession.shared.data(from: url)
            guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let payload = root["data"] as? [String: Any],
                  let native = payload["native"] as? [String: Any],
                  let version = native["version"] as? String,
                  let file = native["file"] as? String,
                  let sha256 = native["sha256"] as? String, sha256.count == 64
            else {
                if interactive { info(L("update.none.title"), L("update.none.body", currentVersion)) }
                return
            }
            guard isNewer(version, than: currentVersion) else {
                if interactive { info(L("update.latest.title"), L("update.latest.body", currentVersion)) }
                return
            }
            // 테스트 훅(개발/E2E 전용): 다이얼로그 없이 즉시 진행 — OMK_COMPANION_* env 훅 계열.
            if ProcessInfo.processInfo.environment["OMK_COMPANION_AUTO_UPDATE"] != "1" {
                let alert = NSAlert()
                alert.messageText = L("update.available.title", version, currentVersion)
                alert.informativeText = L("update.available.body")
                alert.addButton(withTitle: L("update.install"))
                alert.addButton(withTitle: L("update.later"))
                NSApp.activate(ignoringOtherApps: true)
                guard alert.runModal() == .alertFirstButtonReturn else { return }
            }

            let dmgURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(file)
            guard let dl = URL(string: "\(backendUrl)/api/desktop/download/\(file)") else {
                throw UpdateError.message(L("update.badDownload"))
            }
            let (tmp, _) = try await URLSession.shared.download(from: dl)
            try? FileManager.default.removeItem(at: dmgURL)
            try FileManager.default.moveItem(at: tmp, to: dmgURL)

            let digest = SHA256.hash(data: try Data(contentsOf: dmgURL))
            let hex = digest.map { String(format: "%02x", $0) }.joined()
            guard hex == sha256.lowercased() else {
                try? FileManager.default.removeItem(at: dmgURL)
                throw UpdateError.message(L("update.checksum"))
            }

            try spawnReplaceScript(dmgPath: dmgURL.path)
            HelperManager.shared.stopHelper()
            NSApp.terminate(nil)
        } catch let e as UpdateError {
            if case let .message(m) = e, interactive { info(L("update.failed.title"), m) }
        } catch {
            if interactive { info(L("update.failed.title"), error.localizedDescription) }
        }
    }

    private func isSecureOrigin(_ s: String) -> Bool {
        guard let u = URL(string: s), let host = u.host else { return false }
        return u.scheme == "https" || host == "localhost" || host == "127.0.0.1"
    }

    /** 단순 수치 버전 비교 (a > b) — 자리수 다르면 부족분 0 취급. */
    private func isNewer(_ a: String, than b: String) -> Bool {
        let pa = a.split(separator: ".").map { Int($0) ?? 0 }
        let pb = b.split(separator: ".").map { Int($0) ?? 0 }
        for i in 0..<max(pa.count, pb.count) {
            let x = i < pa.count ? pa[i] : 0, y = i < pb.count ? pb[i] : 0
            if x != y { return x > y }
        }
        return false
    }

    /** 교체 스크립트가 실패 시 띄울 문구 — 앱이 다국어로 만들어 env 로 넘긴다(스크립트에 문자열을 박지 않음). */
    private static let scriptMessages: [String: String] = [
        "OMK_UPD_FAIL_TITLE": "update.script.failTitle",
        "OMK_UPD_LOG": "update.script.log",
        "OMK_UPD_MOUNT": "update.script.mount",
        "OMK_UPD_NO_APP": "update.script.noApp",
        "OMK_UPD_COPY": "update.script.copy",
        "OMK_UPD_REPLACE": "update.script.replace",
        "OMK_UPD_ROLLBACK": "update.script.rollback",
        "OMK_UPD_REOPEN": "update.script.reopen",
    ]

    /** 교체 스크립트 — Electron updater.js 의 검증된 시퀀스 이식 (스테이징→스왑→롤백·오류 시 osascript 경고). */
    private func spawnReplaceScript(dmgPath: String) throws {
        guard let appPath = Bundle.main.bundlePath.hasSuffix(".app") ? Bundle.main.bundlePath : nil else {
            throw UpdateError.message(L("update.noBundle"))
        }
        let ts = Int(Date().timeIntervalSince1970)
        let logPath = NSTemporaryDirectory() + "openmake-companion-update-\(ts).log"
        let script = """
        #!/bin/bash
        exec >>"\(logPath)" 2>&1
        set -u
        APP="\(appPath)"
        DMG="\(dmgPath)"

        fail() {
          echo "FAIL: $1"
          /usr/bin/osascript -e "display alert \\"$OMK_UPD_FAIL_TITLE\\" message \\"$1\\n\\n$OMK_UPD_LOG: \(logPath)\\" as critical" || true
          [ -d "$APP" ] && open -a "$APP" || true
          rm -f "$DMG"
          exit 1
        }

        sleep 2
        MNT=$(hdiutil attach -nobrowse "$DMG" | grep -o '/Volumes/.*' | tail -1)
        [ -n "$MNT" ] || fail "$OMK_UPD_MOUNT"
        [ -d "$MNT/OpenMake Companion.app" ] || { hdiutil detach "$MNT" -quiet; fail "$OMK_UPD_NO_APP"; }

        STAGE="$APP.new"
        rm -rf "$STAGE"
        if ! ditto "$MNT/OpenMake Companion.app" "$STAGE"; then
          rm -rf "$STAGE"; hdiutil detach "$MNT" -quiet
          fail "$OMK_UPD_COPY"
        fi
        hdiutil detach "$MNT" -quiet

        BACKUP="$APP.old"
        rm -rf "$BACKUP"
        mv "$APP" "$BACKUP" || { rm -rf "$STAGE"; fail "$OMK_UPD_REPLACE"; }
        if ! mv "$STAGE" "$APP"; then
          mv "$BACKUP" "$APP" 2>/dev/null
          fail "$OMK_UPD_ROLLBACK"
        fi
        rm -rf "$BACKUP"

        xattr -dr com.apple.quarantine "$APP" 2>/dev/null
        rm -f "$DMG"
        open -a "$APP" || fail "$OMK_UPD_REOPEN"
        echo "OK: updated $APP"
        rm -f "$0"
        """
        let shPath = NSTemporaryDirectory() + "openmake-companion-update-\(ts).sh"
        try script.write(toFile: shPath, atomically: true, encoding: .utf8)
        try FileManager.default.setAttributes([.posixPermissions: 0o755], ofItemAtPath: shPath)
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/bin/bash")
        p.arguments = [shPath]
        var env = ProcessInfo.processInfo.environment
        for (name, key) in Self.scriptMessages { env[name] = L(key) }
        p.environment = env
        p.standardOutput = FileHandle.nullDevice
        p.standardError = FileHandle.nullDevice
        try p.run() // detached 성격 — 부모 종료 후에도 계속 실행됨(스크립트가 sleep 후 교체)
    }

    private func info(_ title: String, _ body: String) {
        let a = NSAlert()
        a.messageText = title
        a.informativeText = body
        NSApp.activate(ignoringOtherApps: true)
        a.runModal()
    }
}

enum UpdateError: Error { case message(String) }
